plugin-server/src/cdp/consumers/cdp-behavioural-events.consumer.ts

```diff
@@ -82,13 +89,36 @@ export class CdpBehaviouralEventsConsumer extends CdpConsumerBase {
             // Track events consumed and matched (absolute numbers)
             let eventsMatched = 0
             const counterUpdates: CounterUpdate[] = []
+            const occurrenceUpdates: OccurrenceUpdate[] = []
 
             const results = await Promise.all(events.map((event) => this.processEvent(event, counterUpdates)))
             eventsMatched = results.reduce((sum, count) => sum + count, 0)
 
-            // Batch write all counter updates
+            // Prepare occurrence updates for all events, deduplicating within the batch
+            const occurrenceSet = new Set<string>()
+            events.forEach((event) => {
+                const key = `${event.teamId}:${event.personId}:${event.filterGlobals.event}`
+                if (!occurrenceSet.has(key)) {
+                    occurrenceSet.add(key)
+                    occurrenceUpdates.push({
+                        teamId: event.teamId,
+                        personId: event.personId,
+                        eventName: event.filterGlobals.event,
+                    })
+                }
+            })
+
+            // Batch write all updates in parallel
+            const writePromises = []
             if (counterUpdates.length > 0 && this.hub.WRITE_BEHAVIOURAL_COUNTERS_TO_CASSANDRA && this.cassandra) {
-                await this.writeBehavioralCounters(counterUpdates)
+                writePromises.push(this.writeBehavioralCounters(counterUpdates))
```

Comment:
Do we want to wrap all this in transactions? What if one fails and throws but the other doesnt? Would we retry (whereby breaking the consistency) ?